### PR TITLE
Save criteria changes to the history

### LINKED ISSUE #1196: Save criteria changes to the history
- add/delete
- achieved/unachieved

### PR DESCRIPTION
**Pull request by LamaEats**
## PR includes

- [x] Refactor
- [x] Feature

## Related issues

Resolve #1196 

## QA Instructions, Screenshots, Recordings
  
image

### CODE REVIEW

**Review by awinogradov — CHANGES_REQUESTED** [review]

File: src/utils/db.ts [hunk 1]
```diff
@@ -118,79 +119,105 @@ export const changeGoalProject = async (id: string, newProjectId: string) => {
...
+        if (subjectToEnumValue(subject) && valuesAsIds) {
+            acc[subject] = {
+                ids: (acc[subject]?.ids ?? []).concat(...allValues),
+                sourceIdx: (acc[subject]?.sourceIdx ?? []).concat(index),
+            };
         }
 
         return acc;
-    }, []);
+    }, {} as Record<keyof HistoryRecordSubject, { ids: string[]; sourceIdx: number[] }>);
```
[the remark this answers is not in the document]
Comment (LamaEats) [reply src/utils/db.ts:153]: TS cannon assign type from reduce fn to empty object as initial arg

File: src/utils/db.ts [hunk 2]
```diff
@@ -118,79 +119,105 @@ export const changeGoalProject = async (id: string, newProjectId: string) => {
...
-        if (subjectToEnumValue(subject)) {
-            acc.push(index);
+    const requestParamsBySubjects = history.reduce<
+        Record<keyof HistoryRecordSubject, { ids: string[]; sourceIdx: number[] }>
+    >((acc, { subject, previousValue, nextValue }, index) => {
+        const allValues = (previousValue ?? '')
+            .split(',')
+            .concat((nextValue ?? '').split(','))
+            .filter(Boolean);
+        const valuesAsIds = allValues.length ? allValues.every((val) => stringIsCiudOrId(val)) : false;
```
[the remark this answers is not in the document]
Comment (LamaEats) [reply src/utils/db.ts:143]: #1312

**Review by awinogradov — COMMENTED** [review]
Still...
image

File: src/components/HistoryRecord/HistoryRecord.tsx [hunk 3]
```diff
@@ -314,3 +343,110 @@ export const HistoryRecordParticipant: React.FC<HistoryChangeProps<Activity & {
...
+    if (to) {
+        return (
+            <HistorySimplifyRecord
+                withPretext={false}
+                to={
+                    <HistoryRecordSimplifyCriteria
+                        title={to.title}
+                        state={isChangeAction && action === 'complete'}
+                        strike={action === 'remove'}
+                        isChangeAction={isChangeAction}
```
Comment (awinogradov) [inline src/components/HistoryRecord/HistoryRecord.tsx:444]: Why not to do this same as on [428](https://github.com/taskany-inc/issues/pull/1294/files#diff-28e7bb86d913c391369eee9c1f6a73620a9b73b57ac0771d78c2eff181931ac4R428) line? It's better in terms of composition.

**Review by awinogradov — APPROVED** [review]

**Pull request merged**